src/mug/tls.gleam

```diff
@@ -0,0 +1,630 @@
+  /// erlang application in [`gleam.toml`](https://gleam.run/writing-gleam/gleam-toml/).
+  SslNotStarted
+
+  // For connect only
+  /// An invalid option was passed
+  Options(opt: Dynamic)
+
+  /// TLS connection failed
+  TlsAlert(alert: TlsAlert, description: String)
+
+  // https://www.erlang.org/doc/man/inet#type-posix
+  Eaddrinuse
+  Eaddrnotavail
+  Eafnosupport
+  Ealready
+  Econnaborted
+  Econnrefused
+  Econnreset
+  Edestaddrreq
+  Ehostdown
+  Ehostunreach
+  Einprogress
+  Eisconn
+  Emsgsize
+  Enetdown
+  Enetunreach
+  Enopkg
+  Enoprotoopt
+  Enotconn
+  Enotty
+  Enotsock
+  Eproto
+  Eprotonosupport
+  Eprototype
+  Esocktnosupport
+  Etimedout
+  Ewouldblock
+  Exbadport
+  Exbadseq
+  Nxdomain
+
+  // https://www.erlang.org/doc/man/file#type-posix
+  Eacces
+  Eagain
+  Ebadf
+  Ebadmsg
+  Ebusy
+  Edeadlk
+  Edeadlock
+  Edquot
+  Eexist
+  Efault
+  Efbig
+  Eftype
+  Eintr
+  Einval
+  Eio
+  Eisdir
+  Eloop
+  Emfile
+  Emlink
+  Emultihop
+  Enametoolong
+  Enfile
+  Enobufs
+  Enodev
+  Enolck
+  Enolink
+  Enoent
+  Enomem
+  Enospc
+  Enosr
+  Enostr
+  Enosys
+  Enotblk
+  Enotdir
+  Enotsup
+  Enxio
+  Eopnotsupp
+  Eoverflow
+  Eperm
+  Epipe
+  Erange
+  Erofs
+  Espipe
+  Esrch
+  Estale
+  Etxtbsy
+  Exdev
+}
+
+pub type TlsConnectionOptions {
+  TlsConnectionOptions(
+    host: String,
+    port: Int,
+    timeout: Int,
+    certificates: Certificates,
+  )
+}
+
+/// The certificates to use 
+pub type Certificates {
+  /// Uses these CA certificates and regular certificates to verify the server's certificate.
+  ///
+  /// The `use_system_cacerts` option makes mug use the system's CA certificates, which are 
+  /// usually set to a group of competent CAs who sign most of the web's certificates.  
+  /// You may specifiy your own CA certificates with the `cacerts` option, and custom
+  /// certificates and their keys with the `certs_keys` option.
+  ///
+  /// Note that specifying a PEM encoded CA certificate file will result in the system CA
+  /// certificates not being used. This is because of how the underlying [erlang implementation](https://www.erlang.org/doc/apps/ssl/ssl.html#t:client_option/0)
+  /// works. System CA certificates are passed as DER-encoded certificates, and DER encoded
+  /// certs override PEM-encoded ones. Therefore, if you wish to use a PEM encoded CA cert along
+  /// with the system's CA, you should decode the PEM into a DER using (`public_key:pem_decode`)[https://www.erlang.org/doc/apps/public_key/public_key.html#pem-api]
+  /// and use the DerEncodedCaCertificates variant instead.
+  Certificates(
+    use_system_cacerts: Bool,
+    cacerts: Option(CaCertificates),
+    certs_keys: List(CertsKeys),
+  )
+  /// Do not verify certificates. While this does allow you to use self-signed certificates,
+  /// it is highly recommended to not skip verification, and add a custom CA instead.
+  NoVerification
```

Comment:
This is not a certificate and the name isn't scary enough. Rename this type and the variants please.

Reply:
I have renamed the type to `VerificationMethod`, but I do not feel that it is necessary to rename `NoVerification` into something dangerous. From the issue discussion I've linked above, and from the things mentioned in the `ssl` documentation, there seems to be an "encouragement" of some sorts to disable verification, instead of using custom self-signed certificates.

Please correct me if I've misunderstood.

Reply:
This will not be merged without this change.

The API must very strongly discourage the programmer from disabling TLS verification. There has been multiple instances of vulnerabilities due to inexperienced developers disabling TLS verification without understanding, and it is our obligation as library creators to prevent that.

Reply:
I see. Fixed.